core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -1013,6 +1015,23 @@ class LogCleanerTest {
       expectedCleanableSize / (expectedCleanSize + expectedCleanableSize).toDouble, 1.0e-6d)
   }
 
+  @Test
+  def testEmptySegmentWillNotBeCreatedAfterCleaning(): Unit = {
+    val cleaner = makeCleaner(Int.MaxValue)
+
+    val logProps = new Properties
+    logProps.put(LogConfig.SegmentBytesProp, 1024: java.lang.Integer)
+
+    val log = makeLog(config = LogConfig.fromProps(logConfig.originals, logProps))
+
+    // append keyed messages to generate four segments
+    while (log.numberOfSegments < 4)
+      log.appendAsLeader(record(key = 0, value = log.logEndOffset.toInt), leaderEpoch = 0)
+
+    cleaner.clean(LogToClean(new TopicPartition("test", 0), log, 0, log.activeSegment.baseOffset))
+    assertEquals("Empty segment(s) should not be created after cleaning.", 2, log.numberOfSegments)
```

Comment:
Not sure if this test exposes the existing issue. Even without this patch, the cleaner will collapse the first 3 segments into one segment (with 1 record in it). To expose this issue, we have to force multiple log segments to be generated in the old logic. One way to do that is to generate log segments that are more than 2 billion apart in offsets.

Reply:
Without this patch, this test will generate 4 segments. The first two segments are empty. The latter ones are non-empty.